File: common/src/analyzer.rs
//! 解析ロジック（CLI/WASM共通）
//!
//! Step1結果からの工種自動判定

use crate::types::{RawImageData, WorkTypeDefinition};
use crate::work_type_defs::DEFAULT_WORK_TYPE_DEFINITIONS;
use std::collections::HashSet;
use thiserror::Error;

/// AI解析関連エラー（パース失敗・tagger空結果）
#[derive(Error, Debug)]
pub enum AnalyzerError {
    /// AIレスポンスのパース失敗（再試行可能かの判断に使う）
    #[error("AI response parse error: {0}")]
    ResponseParse(String),

    /// photo-tagger結果が空
    #[error("Photo tagger returned no results")]
    TaggerEmpty,
}

/// Step1結果から工種を自動判定（デフォルト定義使用）
pub fn detect_work_types(raw_data: &[RawImageData]) -> Vec<String> {
    detect_work_types_with(raw_data, DEFAULT_WORK_TYPE_DEFINITIONS)
}

/// Step1結果から工種を自動判定（カスタム定義使用）
pub fn detect_work_types_with(raw_data: &[RawImageData], definitions: &[WorkTypeDefinition]) -> Vec<String> {
    let mut types = HashSet::new();

    for r in raw_data {
        let cat = r.photo_category.to_lowercase();
        let text = r.detected_text.to_lowercase();
        let scene = r.scene_description.to_lowercase();

        for def in definitions {
            let matched = def.category_keywords.iter().any(|kw| cat.contains(&kw.to_lowercase()))
                || def.text_keywords.iter().any(|kw| text.contains(&kw.to_lowercase()))
                || def.scene_keywords.iter().any(|kw| scene.contains(&kw.to_lowercase()));
            if matched {
                types.insert(def.name.to_string());
            }
        }
    }

    types.into_iter().collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_detect_work_types_pavement() {
        // 舗装工検出
        let raw_data = vec![
            RawImageData {
                file_name: "temp1.jpg".to_string(),
                photo_category: "到着温度".to_string(),
                scene_description: "アスファルト舗装".to_string(),
                ..Default::default()
            },
        ];

        let types = detect_work_types(&raw_data);
        assert!(types.contains(&"舗装工".to_string()));
    }

    #[test]
    fn test_detect_work_types_marking() {
        // 区画線工検出
        let raw_data = vec![
            RawImageData {
                file_name: "line1.jpg".to_string(),
                detected_text: "区画線施工".to_string(),
                scene_description: "白線を引いている".to_string(),
                ..Default::default()
            },
        ];

        let types = detect_work_types(&raw_data);
        assert!(types.contains(&"区画線工".to_string()));
    }

    #[test]
    fn test_detect_work_types_multiple() {
        // 複数工種検出
        let raw_data = vec![
            RawImageData {
                file_name: "temp1.jpg".to_string(),
                photo_category: "転圧状況".to_string(),
                scene_description: "ローラーで転圧".to_string(),
                ..Default::default()
            },
            RawImageData {
                file_name: "line1.jpg".to_string(),
                scene_description: "区画線の白線".to_string(),
                ..Default::default()
            },
            RawImageData {
                file_name: "demolish1.jpg".to_string(),
                photo_category: "取壊し状況".to_string(),
                scene_description: "解体作業".to_string(),
                ..Default::default()
            },
        ];

        let types = detect_work_types(&raw_data);
        assert!(types.contains(&"舗装工".to_string()));
        assert!(types.contains(&"区画線工".to_string()));
        assert!(types.contains(&"構造物撤去工".to_string()));
        assert_eq!(types.len(), 3);
    }

    #[test]
    fn test_detect_work_types_empty() {
        // 該当なし
        let raw_data = vec![
            RawImageData {
                file_name: "other.jpg".to_string(),
                photo_category: "その他".to_string(),
                scene_description: "風景写真".to_string(),
                ..Default::default()
            },
        ];

        let types = detect_work_types(&raw_data);
        assert!(types.is_empty());
    }

    #[test]
    fn test_detect_work_types_drainage() {
        // 排水構造物工検出
        let raw_data = vec![
            RawImageData {
                file_name: "drain1.jpg".to_string(),
                detected_text: "側溝設置".to_string(),
                scene_description: "マンホール".to_string(),
                ..Default::default()
            },
        ];

        let types = detect_work_types(&raw_data);
        assert!(types.contains(&"排水構造物工".to_string()));
    }

    #[test]
    fn test_detect_work_types_manhole() {
        // 人孔改良工検出
        let raw_data = vec![
            RawImageData {
                file_name: "manhole1.jpg".to_string(),
                detected_text: "人孔改良工事".to_string(),
                ..Default::default()
            },
        ];

        let types = detect_work_types(&raw_data);
        assert!(types.contains(&"人孔改良工".to_string()));
    }

    #[test]
    fn test_detect_work_types_earthwork() {
        // 道路土工検出
        let raw_data = vec![
            RawImageData {
                file_name: "earth1.jpg".to_string(),
                photo_category: "掘削状況".to_string(),
                scene_description: "バックホウで掘削".to_string(),
                ..Default::default()
            },
        ];

        let types = detect_work_types(&raw_data);
        assert!(types.contains(&"道路土工".to_string()));
    }

}

File: photo-ai-go/internal/analyzer/analyzer.go
// Package analyzer implements work-type automatic detection from Step1 image data.
//
// Ports common/src/analyzer.rs (detect_work_types, detect_work_types_with)
// and common/src/work_type_defs.rs (DEFAULT_WORK_TYPE_DEFINITIONS) from Rust.
//
// Issue #165 – Stream G.
package analyzer

import (
	"errors"
	"fmt"
	"sort"
	"strings"

	"github.com/YuujiKamura/photo-ai-go/internal/ai"
)

// AnalyzerError variants mirror the Rust AnalyzerError enum.
var (
	// ErrTaggerEmpty is returned when the photo-tagger produced no results.
	// Mirrors: AnalyzerError::TaggerEmpty
	ErrTaggerEmpty = errors.New("Photo tagger returned no results")
)

// ResponseParseError mirrors AnalyzerError::ResponseParse(String).
// Use fmt.Errorf("AI response parse error: %s%w", detail, ResponseParseError)
// to create wrapped instances, or use NewResponseParseError.
type ResponseParseError struct {
	Detail string
}

func (e *ResponseParseError) Error() string {
	return fmt.Sprintf("AI response parse error: %s", e.Detail)
}

// NewResponseParseError creates a ResponseParseError, mirroring
// AnalyzerError::ResponseParse(detail).
func NewResponseParseError(detail string) error {
	return &ResponseParseError{Detail: detail}
}

// WorkTypeDefinition holds keyword sets for a single work type.
// Mirrors common/src/types.rs WorkTypeDefinition (with owned strings for Go).
//
// NOTE: Rust uses &'static str slices; Go uses []string. Semantics identical.
type WorkTypeDefinition struct {
	// Name is the work-type label (e.g. "舗装工").
	Name string
	// CategoryKeywords match against RawImageData.PhotoCategory (case-insensitive).
	CategoryKeywords []string
	// TextKeywords match against RawImageData.DetectedText (case-insensitive).
	TextKeywords []string
	// SceneKeywords match against RawImageData.SceneDescription (case-insensitive).
	SceneKeywords []string
}

// DefaultWorkTypeDefinitions is the Go equivalent of DEFAULT_WORK_TYPE_DEFINITIONS
// from common/src/work_type_defs.rs. Values are reproduced verbatim.
var DefaultWorkTypeDefinitions = []WorkTypeDefinition{
	{
		Name:             "舗装工",
		CategoryKeywords: []string{"温度", "転圧", "舗設", "敷均し", "乳剤", "路盤"},
		TextKeywords:     []string{"アスファルト"},
		SceneKeywords:    []string{"アスファルト", "フィニッシャー", "ローラー"},
	},
	{
		Name:             "区画線工",
		CategoryKeywords: []string{"区画線"},
		TextKeywords:     []string{"区画線", "ライン"},
		SceneKeywords:    []string{"白線", "区画線"},
	},
	{
		Name:             "構造物撤去工",
		CategoryKeywords: []string{"取壊し"},
		TextKeywords:     []string{"撤去", "取壊"},
		SceneKeywords:    []string{"解体", "撤去"},
	},
	{
		Name:             "道路土工",
		CategoryKeywords: []string{"掘削", "路床"},
		TextKeywords:     []string{"掘削"},
		SceneKeywords:    []string{"掘削", "バックホウ"},
	},
	{
		Name:             "排水構造物工",
		CategoryKeywords: []string{},
		TextKeywords:     []string{"側溝", "集水", "人孔"},
		SceneKeywords:    []string{"側溝", "マンホール"},
	},
	{
		Name:             "人孔改良工",
		CategoryKeywords: []string{},
		TextKeywords:     []string{"人孔改良", "マンホール蓋"},
		SceneKeywords:    []string{},
	},
}

// DetectWorkTypes detects work types from Step1 image data using the default
// keyword definitions.
//
// Ports: pub fn detect_work_types(raw_data: &[RawImageData]) -> Vec<String>
//
// Input uses ai.RawImageData (internal/ai package) which mirrors the Rust
// RawImageData struct field-for-field.
//
// The returned slice is sorted for deterministic output (Rust returns
// HashSet::into_iter which is non-deterministic; we sort to compensate).
func DetectWorkTypes(rawData []ai.RawImageData) []string {
	return DetectWorkTypesWith(rawData, DefaultWorkTypeDefinitions)
}

// DetectWorkTypesWith detects work types using a caller-supplied definition list.
//
// Ports: pub fn detect_work_types_with(raw_data: &[RawImageData], definitions: &[WorkTypeDefinition]) -> Vec<String>
//
// Behavior is identical to the Rust version:
//   - Each RawImageData field is lowercased before matching.
//   - A definition matches if ANY keyword in ANY of its three keyword slices
//     appears as a substring of the corresponding (lowercased) field.
//   - Matched work-type names are deduplicated via a set (map[string]struct{}).
//
// The returned slice is sorted lexicographically for deterministic ordering.
// (Rust uses HashSet whose iteration order is undefined; sorting makes tests stable.)
func DetectWorkTypesWith(rawData []ai.RawImageData, definitions []WorkTypeDefinition) []string {
	seen := make(map[string]struct{})

	for _, r := range rawData {
		cat := strings.ToLower(r.PhotoCategory)
		text := strings.ToLower(r.DetectedText)
		scene := strings.ToLower(r.SceneDescription)

		for _, def := range definitions {
			matched := containsAny(cat, def.CategoryKeywords) ||
				containsAny(text, def.TextKeywords) ||
				containsAny(scene, def.SceneKeywords)
			if matched {
				seen[def.Name] = struct{}{}
			}
		}
	}

	result := make([]string, 0, len(seen))
	for name := range seen {
		result = append(result, name)
	}
	sort.Strings(result)
	return result
}

// containsAny reports whether s contains any of the given keywords as a
// substring (case-sensitive; callers must pre-lowercase both sides).
// Mirrors: def.category_keywords.iter().any(|kw| cat.contains(&kw.to_lowercase()))
func containsAny(s string, keywords []string) bool {
	for _, kw := range keywords {
		if strings.Contains(s, strings.ToLower(kw)) {
			return true
		}
	}
	return false
}
